services/queue/src/artifacts.js

```diff
@@ -358,6 +417,102 @@ builder.declare({
   }
 });
 
+builder.declare({
+  method: 'put',
+  route: '/task/:taskId/runs/:runId/artifacts/:name(*)',
+  name: 'finishArtifact',
+  stability: APIBuilder.stability.experimental,
+  category: 'Artifacts',
+  scopes: 'queue:create-artifact:<taskId>/<runId>',
+  input: 'finish-artifact-request.json#',
+  title: 'Finish Artifact',
+  description: [
+    'This endpoint marks an artifact as present for the given task, and',
+    'should be called when the artifact data is fully uploaded.',
+    '',
+    'The storage types `reference`, `link`, and `error` do not need to',
+    'be finished, as they are finished immediately by `createArtifact`.',
+    'The storage type `s3` does not support this functionality and cannot',
+    'be finished.  In all such cases, calling this method is an input error',
+    '(400).',
+  ].join('\n'),
+}, async function(req, res) {
+  const taskId = req.params.taskId;
+  const runId = parseInt(req.params.runId, 10);
+  const name = req.params.name;
+  const { uploadId } = req.body;
+
+  const [[artifact_row], task] = await Promise.all([
+    this.db.fns.get_queue_artifact(taskId, runId, name),
+    Task.get(this.db, taskId),
+  ]);
+
+  if (!artifact_row || !task || !task.runs[runId]) {
+    return res.reportError(
+      'ResourceNotFound',
+      'Artifact with taskId `{{taskId}}`, runId {{runId}}, name `{{name}}` not found',
+      { taskId, runId, name },
+    );
+  }
+
+  const artifact = artifactUtils.fromDb(artifact_row);
+  const run = task.runs[runId]; // checked to exist above
+
+  const publishCreatedMessage = async () => {
+    await this.publisher.artifactCreated({
+      status: task.status(),
+      artifact: artifactUtils.serialize(artifact),
+      workerGroup: run.workerGroup,
+      workerId: run.workerId,
+      runId,
+    }, task.routes);
+  };
+
+  // if this is of a type that is already finished, that's not allowed
+  if (artifact.storageType !== 'object') {
```

Comment:
The API description lists the artifact types that do not support the finish method, rather than the one that does, but the check here is against the one that does ('object'). Perhaps the API description should instead just say that this is for object artifacts only, to match the code here. I also see that the description is helpful to understand _why_ reference, link and error do not have an associated finish call, so I'm also ok to leave the description as it is. I'll let you decide!

Reply:
Good point -- I switched this around late in the game (initially I allowed finishing any artifact, and just made it a no-op for most artifact types, but later decided that was too much flexibility).  I'll make a constant defining what does and does not require finishing.